README.md
# fluent-logger for Node.js

fluent-logger implementation for Node.js inspired by [fluent-logger-python].

[![NPM](https://nodei.co/npm/fluent-logger.png?downloads=true&downloadRank=true)](https://nodei.co/npm/fluent-logger/)

[![Build Status](https://secure.travis-ci.org/fluent/fluent-logger-node.png?branch=master,develop)](http://travis-ci.org/fluent/fluent-logger-node)

## Install

    $ npm install fluent-logger

## Prerequistes

Fluent daemon should listen on TCP port.

Simple configuration is following:

```aconf
<source>
  @type forward
  port 24224
</source>

<match **.*>
  @type stdout
</match>
```

## Usage

### Send an event record to Fluentd

Singleton style

```js
var logger = require('fluent-logger')
// The 2nd argument can be omitted. Here is a default value for options.
logger.configure('tag_prefix', {
   host: 'localhost',
   port: 24224,
   timeout: 3.0,
   reconnectInterval: 600000 // 10 minutes
});

// send an event record with 'tag.label'
logger.emit('label', {record: 'this is a log'});
```

Instance style

```js
var logger = require('fluent-logger').createFluentSender('tag_prefix', {
   host: 'localhost',
   port: 24224,
   timeout: 3.0,
   reconnectInterval: 600000 // 10 minutes
});
```

The emit method has following signature

```js
.emit([label string], <record object>, [timestamp number/date], [callback function])
```

Where only the `record` argument is required. If the label is set it will be
appended to the configured tag.

### Disable automatic reconnect 
Both Singleton and Instance style can disable automatic reconnect allowing the user to handle reconnect himself
```js
logger.configure('tag_prefix', {
   host: 'localhost',
   port: 24224,
   timeout: 3.0,
   enableReconnect: false // defaults to true
});
```

### Shared key authentication

Logger configuration:

```js
var logger = require('fluent-logger').createFluentSender('dummy', {
  host: 'localhost',
  port: 24224,
  timeout: 3.0,
  reconnectInterval: 600000, // 10 minutes
  security: {
    clientHostname: "client.localdomain",
    sharedKey: "secure_communication_is_awesome"
  }
});
logger.emit('debug', { message: 'This is a message' });
```

Server configuration:

```aconf
<source>
  @type forward
  port 24224
  <security>
    self_hostname input.testing.local
    shared_key secure_communication_is_awesome
  </security>
</source>

<match dummy.*>
  @type stdout
</match>
```

See also [Fluentd](https://github.com/fluent/fluentd) examples.

### TLS/SSL encryption

Logger configuration:

```js
var logger = require('fluent-logger').createFluentSender('dummy', {
  host: 'localhost',
  port: 24224,
  timeout: 3.0,
  reconnectInterval: 600000, // 10 minutes
  security: {
    clientHostname: "client.localdomain",
    sharedKey: "secure_communication_is_awesome"
  },
  tls: true,
  tlsOptions: {
    ca: fs.readFileSync('/path/to/ca_cert.pem')
  }
});
logger.emit('debug', { message: 'This is a message' });
```

Server configuration:

```aconf
<source>
  @type forward
  port 24224
  <transport tls>
    ca_cert_path /path/to/ca_cert.pem
    ca_private_key_path /path/to/ca_key.pem
    ca_private_key_passphrase very_secret_passphrase
  </transport>
  <security>
    self_hostname input.testing.local
    shared_key secure_communication_is_awesome
  </security>
</source>

<match dummy.*>
  @type stdout
</match>
```

FYI: You can generate certificates using fluent-ca-generate command since Fluentd 1.1.0.

See also [How to enable TLS/SSL encryption](https://docs.fluentd.org/input/forward#how-to-enable-tls-encryption).

### Mutual TLS Authentication

Logger configuration:

```js
var logger = require('fluent-logger').createFluentSender('dummy', {
  host: 'localhost',
  port: 24224,
  timeout: 3.0,
  reconnectInterval: 600000, // 10 minutes
  security: {
    clientHostname: "client.localdomain",
    sharedKey: "secure_communication_is_awesome"
  },
  tls: true,
  tlsOptions: {
    ca: fs.readFileSync('/path/to/ca_cert.pem'),
    cert: fs.readFileSync('/path/to/client-cert.pem'),
    key: fs.readFileSync('/path/to/client-key.pem'),
    passphrase: 'very-secret'
  }
});
logger.emit('debug', { message: 'This is a message' });
```

Server configuration:

```aconf
<source>
  @type forward
  port 24224
  <transport tls>
    ca_path /path/to/ca-cert.pem
    cert_path /path/to/server-cert.pem
    private_key_path /path/to/server-key.pem
    private_key_passphrase very_secret_passphrase
    client_cert_auth true
  </transport>
  <security>
    self_hostname input.testing.local
    shared_key secure_communication_is_awesome
  </security>
</source>

<match dummy.*>
  @type stdout
</match>
```

### EventTime support

We can also specify [EventTime](https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#eventtime-ext-format) as timestamp.

```js
var FluentLogger = require('fluent-logger');
var EventTime = FluentLogger.EventTime;
var logger = FluentLogger.createFluentSender('tag_prefix', {
var eventTime = new EventTime(1489547207, 745003500); // 2017-03-15 12:06:47 +0900
logger.emit('tag', { message: 'This is a message' }, eventTime);
```

### Events

* `connect` : Handle [net.Socket Event: connect](https://nodejs.org/api/net.html#net_event_connect)
* `error` : Handle [net.Socket Event: error](https://nodejs.org/api/net.html#net_event_error_1)

```js
var logger = require('fluent-logger').createFluentSender('tag_prefix', {
   host: 'localhost',
   port: 24224,
   timeout: 3.0,
   reconnectInterval: 600000 // 10 minutes
});
logger.on('error', (error) => {
  console.log(error);
});
logger.on('connect', () => {
  console.log('connected!');
});
```

## Logging Library Support

### log4js

Use [log4js-fluent-appender](https://www.npmjs.com/package/log4js-fluent-appender).

### winston

Before using [winston](https://github.com/winstonjs/winston) support, you should install it IN YOUR APPLICATION.

```js
var winston = require('winston');
var config = {
  host: 'localhost',
  port: 24224,
  timeout: 3.0,
  requireAckResponse: true // Add this option to wait response from Fluentd certainly
};
var fluentTransport = require('fluent-logger').support.winstonTransport();
var fluent = new fluentTransport('mytag', config);
var logger = winston.createLogger({
  transports: [fluent, new (winston.transports.Console)()]
});

logger.on('flush', () => {
  console.log("flush");
})

logger.on('finish', () => {
  console.log("finish");
  fluent.sender.end("end", {}, () => {})
});

logger.log('info', 'this log record is sent to fluent daemon');
logger.info('this log record is sent to fluent daemon');
logger.info('end of log message');
logger.end();
```

**NOTE** If you use `winston@2`, you can use `fluent-logger@2.7.0` or earlier. If you use `winston@3`, you can use `fluent-logger@2.8` or later.

### stream

Several libraries use stream as output.

```js
'use strict';
const Console = require('console').Console;
var sender = require('fluent-logger').createFluentSender('tag_prefix', {
   host: 'localhost',
   port: 24224,
   timeout: 3.0,
   reconnectInterval: 600000 // 10 minutes
});
var logger = new Console(sender.toStream('stdout'), sender.toStream('stderr'));
logger.log('this log record is sent to fluent daemon');
setTimeout(()=> sender.end(), 5000);
```


## Options

**tag_prefix**

The tag prefix string.
You can specify `null` when you use `FluentSender` directly.
In this case, you must specify `label` when you call `emit`.

**host**

The hostname. Default value = 'localhost'.

See [socket.connect][1]

**port**

The port to listen to. Default value = 24224.

See [socket.connect][1]

**path**

The path to your Unix Domain Socket.
If you set `path` then fluent-logger ignores `host` and `port`.

See [socket.connect][1]

**timeout**

Set the socket to timetout after `timeout` milliseconds of inactivity
on the socket.

See [socket.setTimeout][2]

**reconnectInterval**

Set the reconnect interval in milliseconds.
If error occurs then reconnect after this interval.

[1]: https://nodejs.org/api/net.html#net_socket_connect_path_connectlistener
[2]: https://nodejs.org/api/net.html#net_socket_settimeout_timeout_callback

**requireAckResponse**

Change the protocol to at-least-once. The logger waits the ack from destination.

**ackResponseTimeout**

This option is used when requireAckResponse is true. The default is 190. This default value is based on popular `tcp_syn_retries`.

**eventMode**

Set [Event Modes](https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#event-modes). This logger supports `Message`, `PackedForward` and `CompressedPackedForward`.
Default is `Message`.

NOTE: We will change default to `PackedForward` and drop `Message` in next major release.

**flushInterval**

Set flush interval in milliseconds. This option has no effect in `Message` mode.
The logger stores emitted events in buffer and flush events for each interval.
Default `100`.

**messageQueueSizeLimit**

Maximum number of messages that can be in queue at the same time. If a new message is received and it overflows the queue then the oldest message will be removed before adding the new item. This option has effect only in `Message` mode. No limit by default.

**security.clientHostname**

Set hostname of this logger. Use this value for hostname based authentication.

**security.sharedKey**

Shared key between client and server.

**security.username**

Set username for user based authentication. Default values is empty string.

**security.password**

Set password for user based authentication. Default values is empty string.

**sendQueueSizeLimit**

Queue size limit in bytes. This option has no effect in `Message` mode. Default is `8 MiB`.

**tls**

Enable TLS for socket.

**tlsOptions**

Options to pass to tls.connect when tls is true.

For more details, see following documents
* [tls.connect()](https://nodejs.org/api/tls.html#tls_tls_connect_options_callback)
* [tls.createSecureContext()](https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options)

**internalLogger**

Set internal logger object for FluentLogger. Use `console` by default.
This logger requires `info` and `error` method.

## Examples
### Winston Integration
An example of integrating with Winston can be found at [`./example/winston`](./example/winston).

You will need Docker Compose to run it. After navigating to `./example/winston`, run `docker-compose up` and then `node index.js`. You should see the Docker logs having an `"it works"` message being output to FluentD.

## License

Apache License, Version 2.0.

[fluent-logger-python]: https://github.com/fluent/fluent-logger-python


## About NodeJS versions

This package is compatible with NodeJS versions >= 6.


CHANGELOG.md
# v3.x

## v3.4.1 - 2019-12-14

### Fixes

* Fix type definition for winston support #163

## v3.4.0 - 2019-11-20

### Improvements

* Gracefully free resources on `.end()` #144
* Update type definitions for TypeScript #145, #147
* Add new option messageQueueSizeLimit #152

### Fixes

* Fix packets on multiple tags get corrput and multiple calls of callbacks on error #155

## v3.3.1 - 2019-02-19

### Fixes

* Set up default error handler for winston #136
* Flush sendQueue after reconnect #136

## v3.3.0 - 2019-01-31

### Improvements

* Improve performance #131

## v3.2.3 - 2019-01-10

### Improvements

* Update type definition according to documentation for security #127

### Fixes

* Fix user based authentication #128 #129

## v3.2.2 - 2018-12-13

### Improvements

* Improve TypeScript definitions #118 #125

## v3.2.1 - 2018-10-19

### Fixes

* Update TypeScript declaration file #114

## v3.2.0 - 2018-09-10

### Improvements

* Allow using a flag to disable automatic reconnect #111

## v3.1.0 - 2018-09-10

### Improvements

* Add sendQueueSizeLimit option #102
* Add TypeScrip declaration file #110

### Fixes

* Fix winston support #108

## v3.0.0 - 2018-07-24

### Improvements

* Drop Node.js 4 support
* Support winston 3

# v2.x

## v2.8.1 - 2018-09-10

### Fixes

* Fix supported winston version

## v2.8.0 - 2018-07-19

### Fixes

* Reset send queue size #99

## v2.7.0 - 2018-05-11

### Improvements

* Support TLS #92

## v2.6.2 - 2018-02-27

### Improvements

* Introduce ESLint #88

### Fixes

* Avoid writing to closed socket #90

## v2.6.1 - 2017-11-16

### Improvements

* Support log level configuration #85

## v2.6.0 - 2017-10-23

### Improvements

* Replace built-in log4js appender to log4js-fluent-appender #82

## v2.5.0 - 2017-10-11

### Improvements

* Support Fluentd v1 protocol handshake #80

## v2.4.4 - 2017-10-10

### Fixes

* Invoke callback function asynchronously in winston transport #81

## v2.4.3 - 2017-09-28

### Fixes

* Fix bugs in v2.4.2 #77

## v2.4.2 - 2017-09-26

This release has bugs.

### Improvements

* Use arrow functions #76

## v2.4.1 - 2017-07-26

### Fixes

* Clear setTimeout when ack is received #68, #69
* Mark log4js 2.0 and later as unsupported #71
* Flush queue step by step #72

## v2.4.0 - 2017-05-30

### Improvements

* Add internal logger #63

### Fixes

* Update supported engines #64

## v2.3.0 - 2017-03-16

### Improvements

* Support EventTime #61
* Support connect event #62

## v2.2.0 - 2016-10-18

### Fixes

* Wrap dataString to record object #54

## v2.1.0 - 2016-09-28

### Improvements

* Support stream #53

## v2.0.1 - 2016-07-27

### Fixes

* Fix CI settings

## v2.0.0 - 2016-07-27

### Improvements

* Support requireAckResponse
* Support setting tag_prefix to `null`
* Improve error handling
* Add winston transport

# v1.x

## v1.2.1 - 2016-07-15
## v1.2.0 - 2016-07-15
## v1.1.1 - 2016-05-09
## v1.1.0 - 2016-02-01
## v1.0.0 - 2016-01-12

# v0.x

Ancient releases.

## v0.5.0 - 2015-12-14
## v0.4.2 - 2015-11-09
## v0.4.1 - 2015-10-30
## v0.4.0 - 2015-10-30
## v0.3.0 - 2015-09-25
## v0.2.5 - 2013-11-24
## v0.2.4 - 2013-06-02
## v0.2.3 - 2013-06-02
## v0.2.2 - 2013-06-02
## v0.2.1 - 2013-01-10
## v0.2.0 - 2013-01-07
## v0.1.0 - 2012-04-20

## v0.0.2 - 2012-02-17

Initial release


lib/sender.js
'use strict';
const EventEmitter = require('events').EventEmitter;
const msgpack = require('msgpack-lite');
const net = require('net');
const stream = require('stream');
const crypto = require('crypto');
const tls = require('tls');
const zlib = require('zlib');
const FluentLoggerError = require('./logger-error');
const EventTime = require('./event-time');

const codec = msgpack.createCodec();
codec.addExtPacker(0x00, EventTime, EventTime.pack);
codec.addExtUnpacker(0x00, EventTime.unpack);

class FluentSender {
  constructor(tag_prefix, options) {
    options = options || {};
    this._eventMode = options.eventMode || 'Message'; // Message, PackedForward, CompressedPackedForward
    if (!/^Message|PackedForward|CompressedPackedForward$/.test(this._eventMode)) {
      throw new FluentLoggerError.ConfigError('Unknown event mode: ' + this._eventMode);
    }
    this.tag_prefix = tag_prefix;
    this.host = options.host || 'localhost';
    this.port = options.port || 24224;
    this.path = options.path;
    this.timeout = options.timeout || 3.0;
    this.tls = !!options.tls;
    this.tlsOptions = options.tlsOptions || {};
    this.enableReconnect = typeof options.enableReconnect === 'boolean' ? options.enableReconnect : true;
    this.reconnectInterval = options.reconnectInterval || 600000; // Default is 10 minutes
    this.requireAckResponse = options.requireAckResponse;
    this.ackResponseTimeout = options.ackResponseTimeout || 190000; // Default is 190 seconds
    this.internalLogger = options.internalLogger || console;
    this._timeResolution = options.milliseconds ? 1 : 1000;
    this._socket = null;
    if (this._eventMode === 'Message') {
      this._sendQueue = []; // queue for items waiting for being sent.
      this._flushInterval = 0;
      this._messageQueueSizeLimit = options.messageQueueSizeLimit || 0;
    } else {
      this._sendQueue = new Map();
      this._flushInterval = options.flushInterval || 100;
      this._sendQueueSizeLimit = options.sendQueueSizeLimit || 8 * 1024 * 1024; // 8MiB
      this._sendQueueSize = 0;
      this._flushSendQueueTimeoutId = null;
      this._compressed = (this._eventMode === 'CompressedPackedForward');
    }
    this._eventEmitter = new EventEmitter();
    // options.security = { clientHostname: "client.localdomain", sharedKey: "very-secret-shared-key" }
    this.security = options.security || {
      clientHostname: null,
      sharedKey: null,
      username: '',
      password: ''
    };
    this.sharedKeySalt = crypto.randomBytes(16).toString('hex');
    // helo, pingpong, established
    this._status = null;
    this._connecting = false;
  }

  emit(/*[label] <data>, [timestamp], [callback] */a0, a1, a2, a3) {
    let label, data, timestamp, callback;
    let timestampOrCallback, cbArg;
    // Label must be string always
    // Data can be almost anything
    // Date can be either timestamp number or Date object
    // Last argument is an optional callback
    if (typeof a0 === 'string') {
      label = a0;
      data = a1;
      timestampOrCallback = a2;
      cbArg = a3;
    } else {
      data = a0;
      timestampOrCallback = a1;
      cbArg = a2;
    }

    if (typeof timestampOrCallback === 'function') {
      callback = timestampOrCallback;
    } else if (timestampOrCallback) {
      timestamp = timestampOrCallback;
      if (typeof cbArg === 'function') {
        callback = cbArg;
      }
    }

    const tag = this._makeTag(label);
    let error;
    let options;
    if (tag === null) {
      options = {
        tag_prefix: this.tag_prefix,
        label: label
      };
      error = new FluentLoggerError.MissingTag('tag is missing', options);
      this._handleEvent('error', error, callback);
      return;
    }
    if (typeof data !== 'object') {
      options = {
        tag_prefix: this.tag_prefix,
        label: label,
        record: data
      };
      error = new FluentLoggerError.DataTypeError('data must be an object', options);
      this._handleEvent('error', error, callback);
      return;
    }

    this._push(tag, timestamp, data, callback);
    this._connect(() => {
      this._flushSendQueue();
    });
  }

  end(label, data, callback) {
    if ((label != null && data != null)) {
      this.emit(label, data, (err) => {
        this._close();
        if (err) {
          this._handleEvent('error', err, callback);
        } else {
          callback && callback();
        }
      });
    } else {
      process.nextTick(() => {
        this._close();
        callback && callback();
      });
    }
  }

  _close() {
    if (this._socket) {
      this._socket.end();
      this._socket = null;
      this._status = null;
    }
  }

  _makeTag(label) {
    let tag = null;
    if (this.tag_prefix && label) {
      tag = `${this.tag_prefix}.${label}`;
    } else if (this.tag_prefix) {
      tag = this.tag_prefix;
    } else if (label) {
      tag = label;
    }
    return tag;
  }

  _makePacketItem(tag, time, data) {
    if (!time || (typeof time !== 'number' && !(time instanceof EventTime))) {
      time = Math.floor((time ? time.getTime() : Date.now()) / this._timeResolution);
    }

    const packet = [tag, time, data];
    const options = {};
    if (this.requireAckResponse) {
      options.chunk = crypto.randomBytes(16).toString('base64');
      packet.push(options);
    }
    return {
      packet: msgpack.encode(packet, { codec: codec }),
      tag: tag,
      time: time,
      data: data,
      options: options
    };
  }

  _makeEventEntry(time, data) {
    if (!time || (typeof time !== 'number' && !(time instanceof EventTime))) {
      time = Math.floor((time ? time.getTime() : Date.now()) / this._timeResolution);
    }

    return msgpack.encode([time, data], { codec: codec });
  }

  _push(tag, time, data, callback) {
    if (this._eventMode === 'Message') {
      // Message mode
      const item = this._makePacketItem(tag, time, data);
      item.callback = callback;
      if (this._messageQueueSizeLimit && this._sendQueue.length === this._messageQueueSizeLimit) {
        this._sendQueue.shift();
      }
      this._sendQueue.push(item);
    } else {
      // PackedForward mode
      const eventEntry = this._makeEventEntry(time, data);
      this._sendQueueSize += eventEntry.length;
      if (this._sendQueue.has(tag)) {
        const eventEntryData = this._sendQueue.get(tag);
        eventEntryData.eventEntries.push(eventEntry);
        eventEntryData.size += eventEntry.length;
        if (callback) eventEntryData.callbacks.push(callback);
      } else {
        const callbacks = callback ? [callback] : [];
        this._sendQueue.set(tag, {
          eventEntries: [eventEntry],
          size: eventEntry.length,
          callbacks: callbacks
        });
      }
    }
  }

  _connect(callback) {
    if (this._connecting) {
      return;
    }

    if (this._socket === null) {
      this._connecting = true;
      this._doConnect(() => {
        this._connecting = false;
        callback();
      });
    } else if (!this._socket.writable) {
      this._disconnect();
      this._connect(callback);
    } else {
      this._connecting = false;
      process.nextTick(callback);
    }
  }

  _doConnect(callback) {
    const addHandlers = () => {
      const errorHandler = (err) => {
        if (this._socket) {
          this._disconnect();
          this._handleEvent('error', err);
        }
      };
      this._socket.on('error', errorHandler);
      this._socket.on('connect', () => {
        this._handleEvent('connect');
      });
      if (this.tls) {
        this._socket.on('tlsClientError', errorHandler);
        this._socket.on('secureConnect', () => {
          this._handleEvent('connect');
        });
      }
    };
    if (!this.tls) {
      this._socket = new net.Socket();
      this._socket.setTimeout(this.timeout);
      addHandlers();
    }
    if (this.path) {
      if (this.tls) {
        this._socket = tls.connect(Object.assign({}, this.tlsOptions, { path: this.path }), () => {
          callback();
        });
        addHandlers();
      } else {
        this._socket.connect(this.path, () => {
          callback();
        });
      }
    } else {
      const postConnect = () => {
        if (this.security.clientHostname && this.security.sharedKey !== null) {
          this._handshake(callback);
        } else {
          this._status = 'established';
          callback();
        }
      };
      if (this.tls) {
        this._socket = tls.connect(Object.assign({}, this.tlsOptions, { host: this.host, port: this.port }), () => {
          postConnect();
        });
        addHandlers();
      } else {
        this._socket.connect(this.port, this.host, () => {
          postConnect();
        });
      }
    }
  }

  _disconnect() {
    this._socket && this._socket.destroy();
    this._socket = null;
    this._status = null;
    this._connecting = false;
  }

  _handshake(callback) {
    if (this._status === 'established') {
      return;
    }
    this._status = 'helo';
    this._socket.once('data', (data) => {
      this._socket.pause();
      const heloStatus = this._checkHelo(data);
      if (!heloStatus.succeeded) {
        this.internalLogger.error('Received invalid HELO message from ' + this._socket.remoteAddress);
        this._disconnect();
        return;
      }
      this._status = 'pingpong';
      this._socket.write(this._generatePing(), () => {
        this._socket.resume();
        this._socket.once('data', (data) => {
          const pongStatus = this._checkPong(data);
          if (!pongStatus.succeeded) {
            this.internalLogger.error(pongStatus.message);
            const error = new FluentLoggerError.HandshakeError(pongStatus.message);
            this._handleEvent('error', error);
            this._disconnect();
            return;
          }
          this._status = 'established';
          this.internalLogger.info('Established');
          callback();
        });
      });
    });
  }

  _flushSendQueue() {
    if (this._flushingSendQueue)
      return;

    this._flushingSendQueue = true;
    this._waitToWrite();
  }

  _waitToWrite() {
    if (!this._socket) {
      this._flushingSendQueue = false;
      return;
    }

    if (this._socket.writable) {
      if (this._eventMode === 'Message') {
        this._doFlushSendQueue();
      } else {
        if (this._sendQueueSize >= this._sendQueueSizeLimit) {
          this._flushSendQueueTimeoutId && clearTimeout(this._flushSendQueueTimeoutId);
          this._doFlushSendQueue();
        } else {
          this._flushSendQueueTimeoutId && clearTimeout(this._flushSendQueueTimeoutId);
          this._flushSendQueueTimeoutId = setTimeout(() => {
            if (!this._socket) {
              this._flushingSendQueue = false;
              return;
            }
            this._doFlushSendQueue();
          }, this._flushInterval);
        }
      }
    } else {
      process.nextTick(() => {
        this._waitToWrite();
      });
    }
  }

  _doFlushSendQueue(timeoutId) {
    if (this._eventMode === 'Message') {
      const item = this._sendQueue.shift();
      if (item === undefined) {
        this._flushingSendQueue = false;
        // nothing written;
        return;
      }
      this._doWrite(item.packet, item.options, timeoutId, [item.callback]);
    } else {
      if (this._sendQueue.size === 0) {
        this._flushingSendQueue = false;
        return;
      }
      const first = this._sendQueue.entries().next().value;
      const tag = first[0];
      const eventEntryData = first[1];
      let entries = Buffer.concat(eventEntryData.eventEntries, eventEntryData.size);
      let size = entries.length;
      this._sendQueue.delete(tag);
      if (this._compressed) {
        entries = zlib.gzipSync(entries);
        size = entries.length;
      }
      const options = {
        chunk: crypto.randomBytes(16).toString('base64'),
        size: size,
        compressed: this._compressed ? 'gzip' : 'text',
        eventEntryDataSize: eventEntryData.size
      };
      const packet = msgpack.encode([tag, entries, options], { codec: codec });
      this._doWrite(packet, options, timeoutId, eventEntryData.callbacks);
    }
  }

  _doWrite(packet, options, timeoutId, callbacks) {
    const sendPacketSize = (options && options.eventEntryDataSize) || this._sendQueueSize;
    this._socket.write(packet, () => {
      if (this.requireAckResponse) {
        this._socket.once('data', (data) => {
          timeoutId && clearTimeout(timeoutId);
          const response = msgpack.decode(data, { codec: codec });
          if (response.ack !== options.chunk) {
            const error = new FluentLoggerError.ResponseError(
              'ack in response and chunk id in sent data are different',
              { ack: response.ack, chunk: options.chunk }
            );
            callbacks.forEach((callback) => {
              this._handleEvent('error', error, callback);
            });
          } else { // no error on ack
            callbacks.forEach((callback) => {
              callback && callback();
            });
          }
          this._sendQueueSize -= sendPacketSize;
          process.nextTick(() => {
            this._waitToWrite();
          });
        });
        timeoutId = setTimeout(() => {
          const error = new FluentLoggerError.ResponseTimeout('ack response timeout');
          callbacks.forEach((callback) => {
            this._handleEvent('error', error, callback);
          });
        }, this.ackResponseTimeout);
      } else {
        this._sendQueueSize -= sendPacketSize;
        callbacks.forEach((callback) => {
          callback && callback();
        });
        process.nextTick(() => {
          this._waitToWrite();
        });
      }
    });
  }

  _handleEvent(signal, data, callback) {
    callback && callback(data);
    if (this._eventEmitter.listenerCount(signal) > 0) {
      this._eventEmitter.emit(signal, data);
    }
  }

  _setupErrorHandler(callback) {
    if (!this.reconnectInterval || !this.enableReconnect) {
      return;
    }
    this.on('error', (error) => {
      this._flushingSendQueue = false;
      this._status = null;
      this.internalLogger.error('Fluentd error', error);
      this.internalLogger.info('Fluentd will reconnect after ' + this.reconnectInterval / 1000 + ' seconds');
      const timeoutId = setTimeout(() => {
        this.internalLogger.info('Fluentd is reconnecting...');
        this._connect(() => {
          this._flushSendQueue();
          this.internalLogger.info('Fluentd reconnection finished!!');
        });
      }, this.reconnectInterval);
      callback && callback(timeoutId);
    });
  }

  _checkHelo(data) {
    // ['HELO', options(hash)]
    this.internalLogger.info('Checking HELO...');
    const message = msgpack.decode(data);
    if (message.length !== 2) {
      return { succeeded: false, message: 'Invalid format for HELO message' };
    }
    if (message[0] !== 'HELO') {
      return { succeeded: false, message: 'Invalid format for HELO message' };
    }
    const options = message[1] || {};
    this.sharedKeyNonce = options['nonce'] || '';
    this.authentication = options['auth'] || '';
    return { succeeded: true };
  }

  _generatePing() {
    // [
    //   'PING',
    //   client_hostname,
    //   shared_key_salt,
    //   sha512_hex(sharedkey_salt + client_hostname + nonce + shared_key),
    //   username || '',
    //   sha512_hex(auth_salt + username + password) || ''
    // ]
    const sharedKeyHexdigest = crypto.createHash('sha512')
      .update(this.sharedKeySalt)
      .update(this.security.clientHostname)
      .update(this.sharedKeyNonce)
      .update(this.security.sharedKey)
      .digest('hex');
    const ping = ['PING', this.security.clientHostname, this.sharedKeySalt, sharedKeyHexdigest];
    if (Buffer.isBuffer(this.authentication) && this.authentication.length !== 0) {
      const passwordHexDigest = crypto.createHash('sha512')
        .update(this.authentication)
        .update(this.security.username || '')
        .update(this.security.password || '')
        .digest('hex');
      ping.push(this.security.username, passwordHexDigest);
    } else {
      ping.push('', '');
    }
    return msgpack.encode(ping);
  }

  _checkPong(data) {
    // [
    //   'PONG',
    //   bool(authentication result),
    //   'reason if authentication failed',
    //   server_hostname,
    //   sha512_hex(salt + server_hostname + nonce + sharedkey)
    // ]
    this.internalLogger.info('Checking PONG...');
    const message = msgpack.decode(data);
    if (message.length !== 5) {
      return false;
    }
    if (message[0] !== 'PONG') {
      return { succeeded: false, message: 'Invalid format for PONG message' };
    }
    const _pong = message[0];
    const authResult = message[1];
    const reason = message[2];
    const hostname = message[3];
    const sharedKeyHexdigest = message[4];
    if (!authResult) {
      return { succeeded: false, message: 'Authentication failed: ' + reason };
    }
    if (hostname === this.security.clientHostname) {
      return { succeeded: false, message: 'Same hostname between input and output: invalid configuration' };
    }
    const clientsideHexdigest = crypto.createHash('sha512')
      .update(this.sharedKeySalt)
      .update(hostname)
      .update(this.sharedKeyNonce)
      .update(this.security.sharedKey)
      .digest('hex');
    if (sharedKeyHexdigest !== clientsideHexdigest) {
      return { succeeded: false, message: 'Sharedkey mismatch' };
    }
    return { succeeded: true };
  }

  toStream(options) {
    if (typeof options === 'string') {
      options = {label: options};
    } else {
      options = options || {};
    }
    const label = options.label;
    if (!label) {
      throw new Error('label is needed');
    }
    const defaultEncoding = options.encoding || 'UTF-8';
    const writable = new stream.Writable();
    let dataString = '';
    writable._write = (chunk, encoding, callback) => {
      const dataArray = chunk.toString(defaultEncoding).split(/\n/);
      const next = () => {
        if (dataArray.length) {
          dataString += dataArray.shift();
        }
        if (!dataArray.length) {
          process.nextTick(callback);
          return;
        }
        this.emit(label, { message: dataString }, (err) => {
          if (err) {
            this._handleEvent('error', err, callback);
            return;
          }
          dataString = '';
          next();
        });
      };
      next();
    };
    return writable;
  }
}

['addListener', 'on', 'once', 'removeListener', 'removeAllListeners', 'setMaxListeners', 'getMaxListeners'].forEach((attr, i) => {
  FluentSender.prototype[attr] = function() {
    return this._eventEmitter[attr].apply(this._eventEmitter, Array.from(arguments));
  };
});

module.exports = FluentSender;
